### PR TITLE
Add NAT Gateway support to boto_vpc

### PR DESCRIPTION
**Pull request by kraney**
### What does this PR do?

Adds support for AWS's new NAT Gateway functionality to boto_vpc
### What issues does this PR fix or reference?

None
### Previous Behavior

No support for NAT Gateway
### New Behavior

API support for this functionality is not available in boto2, so this creates a new dependency in boto_vpc on boto3. For the best possible backwards compatibility, existing code that makes use of boto 2 has been disturbed as little as possible. Only new functionality relies on boto 3. The downside of this is that boto_vpc actually depends on both boto2 and boto3 now.

This adds nat_gateway_exists, describe_nat_gateways, create_nat_gateway, and delete_nat_gateway methods to the boto_vpc module. It also adds nat_gateway_present and nat_gateway_absent methods to the boto_vpc state.

Also in support of this, the create_route method has been extended to support nat_gateway_id, nat_gateway_subnet_id, and nat_gateway_subnet_name as arguments for setting up routes via the NAT. Also, boto_vpc.present has been enhanced to add a name tag to the default route for the VPC, so it becomes straightforward to add a route via the NAT to the default routing table for the VPC. (Without this, it's hard to identify the correct route table to use to do this.)

The method describe_route_table has an issue where, if the arguments specified match more than one route table, the results returned are potentially an undefined mix of values from the set of possible results, returned as a single item.

To get past this without breaking backwards compatibility, the describe_route_tables method has been added which is extremely similar, but returns a list of results and therefore works when there's more than one match. It also supports querying by vpc ID This version also relies on boto3 and includes results for NAT Gateways. The older describe_route_table method has not been modified to support NAT gateways, and has been deprecated.
### Tests written?
- [x] Yes
- [ ] No

### CODE REVIEW

**Comment by kraney** [comment]
No space left on device. i don't know how to kick off a fresh build.

File: salt/modules/boto_vpc.py [hunk 1]
```diff
@@ -121,7 +130,15 @@ def __virtual__():
     elif _LooseVersion(boto.__version__) < _LooseVersion(required_boto_version):
         return (False, 'The boto_vpc module could not be loaded: boto library is not required version 2.8.0')
     else:
-
+        return True
```
Comment (rallytime) [inline salt/modules/boto_vpc.py]: If we return in this `else`, the remainder of your code is unreachable.

File: salt/modules/boto_vpc.py [hunk 2]
```diff
@@ -1108,6 +1131,193 @@ def delete_internet_gateway(internet_gateway_id=None,
...
+        return False
+
+
+def nat_gateway_exists(nat_gateway_id=None, subnet_id=None, subnet_name=None,
+                       vpc_id=None, vpc_name=None,
+                       states=('pending', 'available'),
+                       region=None, key=None, keyid=None, profile=None):
+    '''
+    Checks if a nat gateway exists.
+
```
Comment (rallytime) [inline salt/modules/boto_vpc.py]: Can you add some notes in the docs here that explain that this function only works with boto3?

Also a `.. versionadded:: Carbon` directive should be added here as well.

File: salt/modules/boto_vpc.py [hunk 3]
```diff
@@ -1108,6 +1131,193 @@ def delete_internet_gateway(internet_gateway_id=None,
...
+                           region=region, key=key, keyid=keyid,
+                           profile=profile))
+
+
+def describe_nat_gateways(nat_gateway_id=None, subnet_id=None, subnet_name=None,
+                       vpc_id=None, vpc_name=None,
+                       states=('pending', 'available'),
+                       region=None, key=None, keyid=None, profile=None):
+    '''
+    Return a description of nat gateways matching the selection criteria
```
Comment (rallytime) [inline salt/modules/boto_vpc.py]: Same doc comment applies here, as well. You should explain the boto3 dependency and add a versionadded tag.

File: salt/modules/boto_vpc.py [hunk 4]
```diff
@@ -1108,6 +1131,193 @@ def delete_internet_gateway(internet_gateway_id=None,
...
+def create_nat_gateway(subnet_id=None,
+                       subnet_name=None, allocation_id=None,
+                       region=None, key=None, keyid=None, profile=None):
+    '''
+    Create a NAT Gateway within an existing subnet. If allocation_id is
+    specified, the elastic IP address it references is associated with the
+    gateway. Otherwise, a new allocation_id is created and used.
+
+    Returns the nat gateway id if the nat gateway was created and
+    returns False if the nat gateway was not created.
```
Comment (rallytime) [inline salt/modules/boto_vpc.py]: This docstring should also explain the boto3 dependency.

File: salt/modules/boto_vpc.py [hunk 5]
```diff
@@ -1108,6 +1131,193 @@ def delete_internet_gateway(internet_gateway_id=None,
...
+
+def delete_nat_gateway(nat_gateway_id,
+                       release_eips=False, region=None,
+                       key=None, keyid=None, profile=None):
+    '''
+    Delete a nat gateway (by id).
+
+    Returns True if the internet gateway was deleted and otherwise False.
+
+    .. versionadded:: 2015.8.0
```
Comment (rallytime) [inline salt/modules/boto_vpc.py]: This should be `Carbon`, not `2015.8.0`.

File: salt/modules/boto_vpc.py [hunk 6]
```diff
@@ -1108,6 +1131,193 @@ def delete_internet_gateway(internet_gateway_id=None,
...
+        return {'created': True, 'id': r.get('NatGateway', {}).get('NatGatewayId')}
+    except BotoServerError as e:
+        return {'created': False, 'error': salt.utils.boto.get_error(e)}
+
+
+def delete_nat_gateway(nat_gateway_id,
+                       release_eips=False, region=None,
+                       key=None, keyid=None, profile=None):
+    '''
+    Delete a nat gateway (by id).
```
Comment (rallytime) [inline salt/modules/boto_vpc.py]: Can you note the boto3 dependency here, too?

File: salt/modules/boto_vpc.py [hunk 7]
```diff
@@ -2117,6 +2360,89 @@ def describe_route_table(route_table_id=None, route_table_name=None,
         return {'error': salt.utils.boto.get_error(e)}
 
 
+def describe_route_tables(route_table_id=None, route_table_name=None,
+                         vpc_id=None,
+                         tags=None, region=None, key=None, keyid=None,
+                         profile=None):
+    '''
+    Given route table properties, return details of all matching route tables.
```
Comment (rallytime) [inline salt/modules/boto_vpc.py]: Boto3 dependency should be noted for this function, as well.

File: salt/states/boto_vpc.py [hunk 8]
```diff
@@ -1081,3 +1095,140 @@ def route_table_absent(name, region=None,
     ret['changes']['new'] = {'route_table': None}
     ret['comment'] = 'Route table {0} deleted.'.format(name)
     return ret
+
+
+def nat_gateway_present(name, subnet_name=None, subnet_id=None,
+                        region=None, key=None, keyid=None, profile=None):
+    '''
+    Ensure a nat gateway exists within the specified subnet
```
Comment (rallytime) [inline salt/states/boto_vpc.py]: This docstring could also use a boto3 dependency note as well as a `.. versionadded:: Carbon` directive.

File: salt/states/boto_vpc.py [hunk 9]
```diff
@@ -1081,3 +1095,140 @@ def route_table_absent(name, region=None,
...
+    inst = r[0]
+    _id = inst.get('NatGatewayId')
+    ret['comment'] = 'Nat gateway {0} present.'.format(_id)
+    return ret
+
+
+def nat_gateway_absent(name=None, subnet_name=None, subnet_id=None,
+                       region=None, key=None, keyid=None, profile=None):
+    '''
+    Ensure the nat gateway in the named subnet is absent.
```
Comment (rallytime) [inline salt/states/boto_vpc.py]: Same here - boto3 dependency and versionadded.

**Comment by rallytime** [comment]
@kraney Thanks again for another great submission. I had a couple of comments that should be addressed (mostly docs) so we're being clear about the boto2 vs boto3 dependencies. Other than that, there was only one problem I found with the `__virtual__` checks you added.

Currently our jenkins server is having some troubles with pull requests, but once we get them back online we can check to make sure your new tests are running properly. (Thank you for writing them.)

**Comment by rallytime** [comment]
Go Go Jenkins!

**Comment by rallytime** [comment]
@kraney It looks like this will no longer merge cleanly. Can you rebase?

**Comment by rallytime** [comment]
@kraney Hrm...something did not go right here in the latest commit you pushed. All of the boto unit tests are failing with this change: https://jenkins.saltstack.com/job/salt-pr-rs-ubuntu14.04-n/10524/#showFailuresLink

Can you take one more look?

**Comment by kraney** [comment]
Ah - the failures happen only if boto3 is not installed.

**Comment by rallytime** [comment]
Ah, good information. We'll see how this next test run goes. We might need to gate the tests a little more, plus QA will need to add the boto3 installation (if it isn't there already) to our Jenkins server testing states.

**Comment by rallytime** [comment]
@jtand I checked to see if the new tests for the were running on Jenkins, and they're not, likely due to a lack of boto3 being installed in the salt-jenkins states. Can you add that to your list of things to add to Jenkins?

@kraney Thank you very much for this change and writing new tests! We'll be sure to get these running on Jenkins, but since you've already had to rebase this a couple of times, and this is in the develop branch, I am going to get this in so development can move forward. If we have any trouble with the tests once they're running on Jenkins's automated runs, we'll be sure to let you know if we run into trouble. Thanks again!

**Comment by kraney** [comment]
Awesome, thanks! You guys are great, a very pleasant open source project to deal with.

**Comment by rallytime** [comment]
:+1:

**Comment by jtand** [comment]
@rallytime Added boto3

**Comment by rallytime** [comment]
Thanks @jtand!

**Pull request merged**